File: javascript_ts/11.qnamaker/src/myAppInsightsQnAMaker.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { TelemetryClient } from 'applicationinsights';
import { TurnContext } from 'botbuilder';
import { QnAMaker, QnAMakerEndpoint, QnAMakerOptions, QnAMakerResult } from 'botbuilder-ai';
import { EventTelemetry } from 'applicationinsights/out/Declarations/Contracts';

/**
 * Custom wrapper around QnAMaker from botbuilder-ai.
 * Sends custom events to Application Insights.
 */
export class MyAppInsightsQnAMaker extends QnAMaker {
    
    public readonly QnAMsgEvent = 'QnAMessage';

    public requestConfig: QnAMakerOptions;
    public logOriginalMessage: boolean = false;
    public logUserName: boolean = false;

    constructor(endpoint: QnAMakerEndpoint, options?: QnAMakerOptions, logOriginalMessage?: boolean, logUserName?: boolean) {
        super(endpoint);
        this.requestConfig = options;
        this.logOriginalMessage = logOriginalMessage;
        this.logUserName = logUserName;
    }

    /**
     * Calls QnA Maker and then sends a custom Event to Application Insights with the answer and question that matched closest with the user's utterance.
     * Sends the top scoring Question and Answer pairing to Application Insights.
     * @param turnContext The TurnContext instance with the necessary information to perform the calls.
     */
    public async getAnswers(turnContext: TurnContext): Promise<QnAMakerResult[]> {
        const { top, scoreThreshold } = this.requestConfig;

        // Call QnAMaker.generateAnswer to retrieve possible Question and Answer pairings for the user's utterance.
        const results: QnAMakerResult[] = await this.generateAnswer(turnContext.activity.text, top, scoreThreshold);

        // Retrieve the reference for the TelemetryClient that was cached for the Turn in TurnContext.turnState via MyAppInsightsMiddleware.
        const telemetryClient: TelemetryClient = turnContext.turnState.get('AppInsightsLoggerMiddleware.AppInsightsContext');
        const telemetryProperties: { [key: string]: string } = {};
        const telemetryMetrics: { [key: string]: number } = {};
        
        const activity = turnContext.activity;

        // Make it so we can correlate our reports with Activity or Conversation.
        telemetryProperties.ActivityId = activity.id;
        if (activity.conversation.id) {
            telemetryProperties.ConversationId = activity.conversation.id;
        }
        
        // For some customers, logging original text name within Application Insights might be an issue.
        if (this.logOriginalMessage && !!activity.text) {
            telemetryProperties.OriginalQuestion = activity.text;
        }
        
        // For some customers, logging user name within Application Insights might be an issue.
        if (this.logUserName && !!activity.from.name) {
            telemetryProperties.Username = activity.from.name;
        }

        // Fill in QnA Results (found or not).
        if (results.length > 0) {
            const queryResult = results[0];
            telemetryProperties.Question = queryResult.questions[0];
            telemetryProperties.Answer = queryResult.answer;
            telemetryMetrics.Score = queryResult.score;
        } else {
            telemetryProperties.Question = 'No QnA Question matched.';
            telemetryProperties.Answer = 'No QnA Question matched.';
        }

        // Finish constructing the event.
        const qnaMsgEvent: EventTelemetry = { name: 'QnAMessage', 
                                              properties: telemetryProperties,
                                              measurements: telemetryMetrics
                                            };

        // Track the event.
        telemetryClient.trackEvent(qnaMsgEvent);

        return results;
    }
}

File: samples/csharp_webapi/22.QnA-with-AppInsights/AppInsights/MyAppInsightsQnaMaker.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.ApplicationInsights;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.AI.QnA;

namespace QnABot.AppInsights
{
    /// <summary>
    /// MyAppInsightsQnARecognizer invokes the QnA Maker and logs some results into Application Insights.
    /// Logs the score, and (optionally) question along with Conversation and ActivityID.
    ///
    /// Customize for specific reporting needs.
    ///
    /// The Custom Event name this logs is "QnAMessage"
    /// See <seealso cref="QnAMaker"/> for additional information.
    /// </summary>
    public class MyAppInsightsQnAMaker : QnAMaker
    {
        public static readonly string QnAMsgEvent = "QnAMessage";

        /// <summary>
        /// Initializes a new instance of the <see cref="MyAppInsightsQnAMaker"/> class.
        /// </summary>
        /// <param name="endpoint">The endpoint of the knowledge base to query.</param>
        /// <param name="options">The options for the QnA Maker knowledge base.</param>
        /// <param name="logUserName">Option to log user name to Application Insights (PII consideration).</param>
        /// <param name="logOriginalMessage">Option to log the original message to Application Insights (PII consideration).</param>
        /// <param name="httpClient">An alternate client with which to talk to QnAMaker.
        /// If null, a default client is used for this instance.</param>
        public MyAppInsightsQnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, bool logUserName = false, bool logOriginalMessage = false, HttpClient httpClient = null)
            : base(endpoint, options, httpClient)
        {
            LogUserName = logUserName;
            LogOriginalMessage = logOriginalMessage;
        }

        public bool LogUserName { get; }

        public bool LogOriginalMessage { get; }

        public new async Task<QueryResult[]> GetAnswersAsync(ITurnContext context)
        {
            // Call QnA Maker
            var queryResults = await base.GetAnswersAsync(context);

            // Find the Application Insights Telemetry Client
            if (queryResults != null && context.TurnState.TryGetValue(MyAppInsightsLoggerMiddleware.AppInsightsServiceKey, out var telemetryClient))
            {
                var telemetryProperties = new Dictionary<string, string>();
                var telemetryMetrics = new Dictionary<string, double>();

                // Make it so we can correlate our reports with Activity or Conversation
                telemetryProperties.Add(MyQnAConstants.ActivityIdProperty, context.Activity.Id);
                var conversationId = context.Activity.Conversation.Id;
                if (!string.IsNullOrEmpty(conversationId))
                {
                    telemetryProperties.Add(MyQnAConstants.ConversationIdProperty, conversationId);
                }

                // For some customers, logging original text name within Application Insights might be an issue.
                var text = context.Activity.Text;
                if (LogOriginalMessage && !string.IsNullOrWhiteSpace(text))
                {
                    telemetryProperties.Add(MyQnAConstants.OriginalQuestionProperty, text);
                }

                // For some customers, logging user name within Application Insights might be an issue.
                var userName = context.Activity.From.Name;
                if (LogUserName && !string.IsNullOrWhiteSpace(userName))
                {
                    telemetryProperties.Add(MyQnAConstants.UsernameProperty, userName);
                }

                // Fill in Qna Results (found or not)
                if (queryResults.Length > 0)
                {
                    var queryResult = queryResults[0];
                    telemetryProperties.Add(MyQnAConstants.QuestionProperty, string.Join(",", queryResult.Questions));
                    telemetryProperties.Add(MyQnAConstants.AnswerProperty, queryResult.Answer);
                    telemetryMetrics.Add(MyQnAConstants.ScoreProperty, (double)queryResult.Score);
                }
                else
                {
                    telemetryProperties.Add(MyQnAConstants.QuestionProperty, "No Qna Question matched");
                    telemetryProperties.Add(MyQnAConstants.AnswerProperty, "No Qna Question matched");
                }

                // Track the event
                ((TelemetryClient)telemetryClient).TrackEvent(QnAMsgEvent, telemetryProperties, telemetryMetrics);
            }

            return queryResults;
        }
    }
}